oeps/oep-0006.rst

```diff
@@ -0,0 +1,224 @@
+
+Some existing XBlocks currently use workarounds for the lack of context-scoped fields, such as storing data in the fields of the root 'course' XBlock. Converting such existing techniques to use context-scoped fields instead would be best achieved with an XBlock data migration API, which currently does not exist and is outside the scope of this OEP (however, see `here`_ for an example implementation).
+
+.. _here: https://github.com/open-craft/problem-builder/blob/c5879dd/problem_builder/v1/xml_changes.py
+
+Details on the current implementation of field scopes in edx-platform can be found in `this gist`_. The `field test XBlock`_ can be used to test how a runtime behaves with different field scopes.
+
+.. _this GIST: https://gist.github.com/bradenmacdonald/b576c59e46af8df9949fa84b3bc7a4b6
+.. _field test XBlock: https://github.com/open-craft/xblock-field-test
+
+Open Questions
+==============
+
+#. How can an XBlock specify whether a context-scoped field should participate
+   in any sort of auto-generated "course preferences" UI (for
+   BlockScope.CONTEXT_TYPE|UserScope.ONE fields) or "course advanced settings"
+   UI (for BlockScope.CONTEXT_TYPE|UserScope.NONE fields)?
+
+   * This may be outside the scope of this OEP, and would best be addressed at
+     the same time as a standard API for definining editor UI options such as
+     ``allow_reset``, ``multiline_editor``, ``list_style``, and
+     ``values_provider`` as implemented in `xblock-utils`_.
+
+#. The current proposal does not support inheritance (i.e. a "context" in the LMS would always be a course, not a "section" or any smaller unit, and field values don't propagate from parent blocks to children), so it is not a replacement for `InheritanceMixin`_. Should it be?
+
+#. Related: Do we need to consider supporting contexts smaller than a course, such as a section or exam? And if so, could that be done in a future chanage, such as adding a non-scope field property to select a context which defaults to "course"?
+
+   * The author's thought on this is that we likely cannot come up with a reasonable proposal to support every possible use case, and it's better to offer simple, flexible, easily understandable support for the main "course" use case now.
+
+#. Should there be a way to define a field such that the runtime will search through a hierarchy of ``BlockScope`` scp[es and use the first value found, such as "check for a ``Scope.settings`` value, then a ``Scope.context_settings`` value"?
+
+.. _xblock-utils: https://github.com/edx/xblock-utils/blob/48fa1a9/xblockutils/studio_editable.py
+.. _InheritanceMixin: https://github.com/edx/edx-platform/blob/46d69eba7ac45/common/lib/xmodule/xmodule/modulestore/inheritance.py#L30
+
+
```

Comment:
I'm not familiar with the OLX architecture goals, but I think something those lines would work - e.g. a `policy/context_fields/[block_type].json` (or .xml) for storing `context_settings` field values. (Or they could be in `[block_type]/fields.json` though not `[block_type]/fields.xml` since 'fields' could theoretically be used as a block ID, and per-usage fields are stored in `[block_type]/[block_id].xml`)